dict.js

```diff
@@ -46,74 +35,123 @@ Dict.prototype.assertString = function (key) {
+                this.dispatchBeforeMapChange(key, undefined);
+            }
+            this.length++;
+            this["$__proto__"] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return true;
         }
-        this.length++;
-        this.store[mangled] = value;
-        if (this.dispatchesMapChanges) {
-            this.dispatchMapChange(key, value);
+    }
+    else {
+        if (key in this.store) { // update
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, this.store[key]);
+            }
+            this.store[key] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return false;
+        } else { // create
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, undefined);
+            }
+            this.length++;
+            this.store[key] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return true;
         }
-        return true;
     }
 };
 
 Dict.prototype.has = function (key) {
     this.assertString(key);
-    var mangled = mangle(key);
-    return mangled in this.store;
+    return (key === "__proto__") ? this.hasOwnProperty("$__proto__") : (key in this.store);
 };
 
 Dict.prototype["delete"] = function (key) {
     this.assertString(key);
-    var mangled = mangle(key);
-    if (mangled in this.store) {
-        if (this.dispatchesMapChanges) {
-            this.dispatchBeforeMapChange(key, this.store[mangled]);
+    if (key === "__proto__") {
+        if (this.hasOwnProperty("$__proto__")) {
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, this["$__proto__"]);
+            }
+            delete this["$__proto__"];
+            this.length--;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, undefined);
+            }
+            return true;
         }
-        delete this.store[mangle(key)];
-        this.length--;
-        if (this.dispatchesMapChanges) {
-            this.dispatchMapChange(key, undefined);
+        return false;
+    }
+    else {
+        if (key in this.store) {
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, this.store[key]);
+            }
+            delete this.store[key];
+            this.length--;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, undefined);
+            }
+            return true;
         }
-        return true;
+        return false;
     }
-    return false;
 };
 
 Dict.prototype.clear = function () {
-    var key, mangled;
-    for (mangled in this.store) {
-        key = unmangle(mangled);
+    var key;
+    delete this["$__proto__"];
```

Comment:
Yes, on that instance. I could add a test of hasOwnProperty to reduce the cases it would happen. I could also reset the property to undefined. Since the set of a new property send:
                this.dispatchBeforeMapChange(key, undefined);
it implies that unset keys have a value of undefined. correct?